addon/components/base/bs-form.js

```diff
@@ -177,6 +189,33 @@ export default Component.extend({
    */
   isSubmitting: false,
 
+  /**
+   * `isSubmitted` is `true` if last submission was successful.
+   * A change to any form element resets it's value to `false`.
+   *
+   * If not using `Components.FormElement`, `resetSubmissionState` action must be triggered on each change to reset
+   * form's submission state.
+   *
+   * @property isSubmitted
+   * @type {Boolean}
+   * @private
+   */
+  isSubmitted: false,
+
+  /**
+   * `isRejected` is `true` if last submission was rejected.
+   * A submission is considered as rejected if form is invalid as well as if `onSubmit` rejects.
```

Comment:
> Maybe isRejected should not be true if an onInvalid method is passed in and that one resolves?

As more as I think about it I'm tending to prefer that implementation. Assuming a form like `<BsForm @onSubmit={{action "save"}} @onInvalid={{action "save"}}>` it should not make any difference if the model is valid or not.

Will wait for your feedback before updating PR accordingly.

Reply:
> Maybe isRejected should not be true if an onInvalid method is passed in and that one resolves?

Agree. 

IMO `isSubmitted` and `isRejected` should only be influenced by the `onSubmit()` action. The result of `onInvalid()` is not used so far (as the submission state is basically something like `isInvalid`, whatever the result of `onInvalid()` is). Are we on the same page?

Reply:
If I got you right you are suggesting to add an additional submission state `isInvalid` to handle that case? Lets maybe put all cases together to not get confused.

Table only lists the property which is `true`. All other ones should be `false`.

| case                                                          | version a        | version b        | version c        |
| ------------------------------------------------------ | ------------------ | ------------------ | ------------------ | 
| form is valid, `onSubmit` resolves             | `isSubmitted` | `isSubmitted` | `isSubmitted` |
| form is valid, `onSubmit` rejects                | `isRejected`  | `isRejected`   | `isRejected `   |
| form is invalid, `onInvalid` is `undefined`  | `isRejected`   | `isRejected`   | `isInvalid`       |
| form is invalid, `onInvalid` resolves           | `isRejected`   | `isSubmitted` | `isInvalid`       |
| form is invalid, `onInvalid` rejects              | `isRejected`   | `isRejected`   | `isInvalid`       |

_version a_ is current implementation in this PR. _version b_ is what I had in mind in my first comment. _version c_ is what you are suggesting if I got you right.

A _version d_ could combine _version b_ and _version c_ meaning that `isInvalid` just reflects validation state on submission and `isRejected` / `isSubmitted` indicates if event handler resolves.

A strong argument for _version b_ is in my opinion that it would allow to ignore validation errors on form submission without introducing any edge cases. That was the main reason from #730.

Reply:
> If I got you right you are suggesting to add an additional submission state isInvalid to handle that case?

Possibly, but not really required I guess, as one could as well just use the validation library's appropriate property!?

> Assuming a form like <BsForm @onSubmit={{action "save"}} @onInvalid={{action "save"}}> it should not make any difference if the model is valid or not.

I assume if `onSubmit` does not return a Promise, this will still make `isSubmitted` true, right? In this case there would still be a difference between `onSubmit` and `onInvalid`, even in your preferred version b. Maybe you can add a table row "form is valid, onSubmit is `undefined`"?

I am still a bit hesitant for this `onInvalid` case. The case is #730 is still quite unusual IMHO. And not sure if the support for this will not cause unexpected behavior for other cases, where a user's `onInvalid` action might (accidentally) return a promise (say you do a `(perform task)`, with some async things handled for the invalid case, like logging, flash messages etc.), but the semantic of that promise is not really that this should be considered a successful form submission (which might cause some inappropriate rendering)

Reply:
> And not sure if the support for this will not cause unexpected behavior for other cases, where a user's onInvalid action might (accidentally) return a promise [...], but the semantic of that promise is not really that this should be considered a successful form submission

That's a very good argument, which I haven't considered yet. Version b requires `onInvalid` method to be absent or to reject / throw in order to get an `isRejected` state. I think you are totally right that no one would expect such a behavior when implementing an action that handles a submission of invalid form.